pithos/plugins/notify.py

```diff
@@ -52,16 +52,15 @@ def send_notification(self, window, *ignore):
             self._app.withdraw_notification(self._app_id)
         else:
             song = window.current_song
+            album = '\n' + song.album if song.album else ''
             # This matches GNOME-Shell's format
             notification = Gio.Notification.new(song.artist)
             # GNOME focuses the application by default, we want to match that behavior elsewhere such as on KDE.
             notification.set_default_action('app.activate')
-            notification.set_body(song.title)
+            notification.set_body(song.title + album)
+
+            icon = Gio.ThemedIcon.new(song.artUrl) if song.artUrl else self._fallback_icon
```

Comment:
I'm confused how this works tbh, sounds like a bug.

We need to download the data and use `Gio.BytesIcon` actually.

Reply:
The `artUrl` property is already a local file URI in the cache. Why would we want to re-download it?

Reply:
@TingPing I've tried with `Gio.BytesIcon` as well but to no avail, so possibly a bug upstream. Can we merge until the upstream bug is resolved?

Reply:
No. As-is this honestly makes 0 sense.

Themed Icons are icons defined here: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html

Using an HTTP url here is nonsense and shouldn't work or be used.

Reply:
The `artUrl` property is just a file:// URI to the local filesystem. It is downloaded into the cache as the playlist populates. Why doesn't it make sense to reuse the same locally cached files? I'm a little confused on this one.

Reply:
@TingPing L39 in this file... `_fallback_icon` has been a `ThemedIcon` for quite some time now.

Reply:
@TingPing I agree in the unorthodox nature but this is the only `Gio.Icon` implementation I could get to work successfully with `Gio.Notification.set_icon()`.

Reply:
Would you prefer to remove the album art in favor of just the fallback icon? This would solve the original issue of notifications not working on the Flatpak version of the app just the same.

Reply:
I would prefer just using BytesIcon. I'm sure it will work.

Reply:
Do you want to look through some of the force-pushed changes that include the `Gio.BytesIcon` attempts? Maybe it's tangential to the same bug or maybe not, but it is not working. I'll play with it more later but it does not feel like it will trend toward the outcome of success.

Reply:
Hmm. You're usage of BytesIcon seems correct.

Reply:
Testing this locally I can confirm `BytesIcon` didn't work. However your `ThemedIcon` changes also do not work here.

GNOME-Shell 42.3 - xdg-desktop-portal 1.12.6

Reply:
So `xdg-desktop-portal` accepts only BytesIcon or ThemedIcon (indeed FileIcon as before couldn't work) - https://github.com/flatpak/xdg-desktop-portal/blob/main/src/notification.c#L281

Reply:
`glib` does nothing except serialize it as-is: https://github.com/GNOME/glib/blob/main/gio/gportalnotificationbackend.c#L67

Reply:
```bash
$ apt list gnome-shell xdg-desktop-portal
Listing... Done
gnome-shell/jammy-updates,now 42.2-0ubuntu0.2 amd64 [installed,automatic]
xdg-desktop-portal/jammy-updates,now 1.14.4-1ubuntu2~22.04.1 amd64 [installed,automatic]
```

I tested that the icon does not show on GNOME but that the notifications do work. (GNOME also has a redundant MPRIS interface so it may not be as much of an issue there.) If this introduces a regression in notifications on GNOME, I can try to workaround that as well.

On other systems, I have Cinnamon 5.4 and Enlightenment 0.24.2 running and can confirm that the notifications do work from the Flatpak installation where they did not previously.

Reply:
I would guess this is a GNOME-Shell bug but tbh I don't see anything off: https://github.com/GNOME/gnome-shell/blob/main/js/ui/notificationDaemon.js

Reply:
This looks interesting: https://github.com/GNOME/gnome-shell/blob/main/js/ui/notificationDaemon.js#L70

It could be a bug in Gio.

Reply:
This codepath is not hit for us.

`xdg-desktop-portal-gtk` calls `org.gtk.Notifications` handled here: https://github.com/GNOME/gnome-shell/blob/main/js/ui/notificationDaemon.js#L462

What you linked is when `org.freedesktop.Notifications` is called.

Reply:
Oh just saw your other comment. On Cinnamon/Enlightenment maybe they just don't handle BytesIcons.

Reply:
Cinnamon is forked from GNOME 3 so I don't expect too much variance there. Enlightenment, I would need to follow up.

Reply:
When I tested on GNOME 42, the notifications worked but they presented without the icons. Is that the main blocker to merging for now?

Reply:
Cinnamon is a very different codebase. They for example do not listen on `org.gtk.Notifications` at all so its behavior is 100% different: https://github.com/linuxmint/cinnamon/blob/master/js/ui/notificationDaemon.js

Anyway yes I'd want it to at least work on GNOME-Shell.

Reply:
> Cinnamon is a very different codebase. They for example do not listen on `org.gtk.Notifications` at all so its behavior is 100% different: https://github.com/linuxmint/cinnamon/blob/master/js/ui/notificationDaemon.js

Fascinating since it is working fine for me without modifying the talk-name configuration.


> Anyway yes I'd want it to at least work on GNOME-Shell.

The notifications worked for me when I tested on GNOME 42 (sans album photo). Can you share some details with what issues you saw on your end? Is there anything in the logs?

Reply:
> Fascinating since it is working fine for me without modifying the talk-name configuration.

pithos -> xdg-desktop-portal -> xdg-desktop-portal-gtk -> if available org.gtk.Notifications, otherwise org.freedesktop.Notifications. No permission required.

> The notifications worked for me when I tested on GNOME 42 (sans album photo)

Sorry if unclear. The missing album art is the only issue.

Reply:
> > The notifications worked for me when I tested on GNOME 42 (sans album photo)
> 
> Sorry if unclear. The missing album art is the only issue.

Can you confirm whether notifications were working on the current stable Flatpak build as well? I believe notifications from Flatpak were entirely broken (after the initial notification on startup if the window was hidden before playback). If notifications were indeed working on GNOME, we can test for that and use the existing code. Also, if it is just the Flatpak version and GNOME, we can test for that as well. I can add comments around the conditionals and tests we introduce until this is cleaned up upstream. I'll also test with both native and Flatpak builds with GNOME on my local installation when I have the time.

Reply:
@TingPing I tested with a local build and Flatpak and it looks like only the Flatpak for GNOME needs changes as the compiled version worked as expected. Please shout if you have observed anything different from this as I'm getting started on separating out that condition now and hope to have an updated PR for review soon.

Reply:
@TingPing Please feel free to test against the latest PR. This should be good to go for all environments. I added a comment to the code for future contributors as well.